packages/shared/src/authz/domain/workspaces/operations.ts

```diff
@@ -3,28 +3,40 @@ import { WorkspaceRoles } from '../../../core/constants.js'
 import { FeatureFlags } from '../../../environment/index.js'
 import { Workspace, WorkspaceSsoProvider, WorkspaceSsoSession } from './types.js'
 import {
+  WorkspaceNoAccessError,
   WorkspaceNotFoundError,
   WorkspaceRoleNotFoundError,
   WorkspaceSsoProviderNotFoundError,
+  WorkspaceSsoSessionNoAccessError,
   WorkspaceSsoSessionNotFoundError
 } from '../authErrors.js'
 
 export type GetWorkspace = (args: {
   workspaceId: string
-}) => Promise<Result<Workspace, typeof WorkspaceNotFoundError>>
+}) => Promise<
+  Result<
+    Workspace,
+    | InstanceType<typeof WorkspaceNotFoundError>
+    | InstanceType<typeof WorkspaceSsoSessionNoAccessError>
+    | InstanceType<typeof WorkspaceNoAccessError>
+  >
+>
 
 export type GetWorkspaceRole = (args: {
   userId: string
   workspaceId: string
-}) => Promise<Result<WorkspaceRoles, typeof WorkspaceRoleNotFoundError>>
+}) => Promise<Result<WorkspaceRoles, InstanceType<typeof WorkspaceRoleNotFoundError>>>
```

Comment:
I made the decision not to surface WorkspaceNotFound/WorkspaceNoAccess/WorkspaceSsoSessionNoAccess errors out of this one, even tho they can pop up from the loader on the frontend. This is because I wanted to avoid the complication of surfacing all of these error types any time anything related to workspaces is loaded on the FE. I think it's better to assume that the developer will make sure to invoke GetWorkspace first, before moving on and checking more granular aspects of the workspace.

Same idea also in Project loaders